x-pack/plugins/apm/public/components/shared/FilterBar/ServiceEnvFilter.tsx

```diff
@@ -0,0 +1,94 @@
+import { EuiFilterGroup, EuiFilterButton, EuiSelect } from '@elastic/eui';
+import React, { useEffect, useRef, useState } from 'react';
+import { useFetcher } from '../../../hooks/useFetcher';
+import { loadServiceEnvironments } from '../../../services/rest/apm/services';
+import { useLocation } from '../../../hooks/useLocation';
+import { useUrlParams } from '../../../hooks/useUrlParams';
+import { history } from '../../../utils/history';
+import { fromQuery, toQuery } from '../Links/url_helpers';
+
+function without<T>(obj: T, key: string): T {
+  const result: { [objKey: string]: unknown } = Object.assign({}, obj);
+  delete result[key];
+  return result as T;
+}
+
+export const ServiceEnvFilter: React.FC = () => {
+  const location = useLocation();
+  const { urlParams } = useUrlParams();
+  const { start, end, kuery, environment } = urlParams;
+  const { data = [] } = useFetcher(
+    () => loadServiceEnvironments({ start, end, kuery, environment: '' }),
+    [start, end, kuery]
+  );
+  const availableEnvironments = ['Any', ...data, 'Not set'];
+  const [selectedValue, setSelectedValue] = useState(availableEnvironments[0]);
+  const isPostEffect = useRef(false);
+
+  function updateEnvironmentUrl(environment?: string) {
+    if (!environment || environment === availableEnvironments[0]) {
+      history.push({
+        ...location,
+        search: fromQuery(without(toQuery(location.search), 'environment'))
```

Comment:
You don't need the `without` helper:
```
search: fromQuery({
  ...toQuery(location.search),
  environment: null
})
```

Reply:
it looks like this still doesn't clean out the param in the URL, it ends up looking like this: `/app/apm#/services?_g=()&environment=`. The only way i found to remove unused params is to omit it entirely. Alternatively we could modify `fromQuery` to automatically omit `null` or `undefined` params.

Reply:
What's the problem with `/app/apm#/services?_g=()&environment=`?

Reply:
i guess nothing is wrong with it, but to a user, it looks like a bug in the url. in any case, if we wanted to clean it up, it shouldn't be handled at this level, but rather within `fromQuery` if we wanted it.